File: src/almost/problem249.hs

{-
Let S = {2, 3, 5, ..., 4999} be the set of prime numbers less than 5000.

Find the number of subsets of S, the sum of whose elements is a prime number. Enter the rightmost 16 digits as your answer.
-}

import qualified Data.Numbers.Primes as P (primes, isPrime)
import qualified Data.MemoCombinators as Memo

{-
1548136 -- desired sum of S
...
8 |
7 |x x   •   •
6 |x x   x   x
5 |x •   •   •
4 |x x   x   x
3 |x •   •   •
2 |• •   •   •
1 |
0 |
  ------------
   2 3   5   7...4999 -- max usable prime (sUB)
-}























sUB :: Integer
sUB = 7 --5000

s :: [Integer]
s = takeWhile (< sUB) P.primes

-- | Number of ways to form `n` as a summation of primes less than `maxPrime`
numFormations :: Integer -> Integer -> Integer
numFormations 0 _ = 1 -- formation is the empty set
numFormations 1 _ = 0 -- no formations possible
numFormations n maxPrime
    | (n < 0) = 0 -- can't sum to a negative number with only positive numbers
    | otherwise = (Memo.memo2 Memo.integral Memo.integral numFormations') n maxPrime

-- | Number of ways to form `n` as a summation of primes less than `maxPrime`
numFormations' :: Integer -> Integer -> Integer
numFormations' n maxPrime
    = sum
    . map (\p -> numFormations (n - p) p)
    $ takeWhile (< maxPrime) P.primes

main :: IO ()
main = do
    putStrLn $ "S = filter prime [2.." ++ show sUB ++ "]"
    putStr "Number of subsets of S that sum to a prime number: "
    print
        . sum -- TODO: fold with truncating +
        . map (\n -> numFormations n (n + 1)) -- `+ 1` because of noninclusiveness
        . takeWhile (<= (sum s)) -- `sum s` is for the subset that is `S`; the maximal subset
        $ P.primes

File: src/almost/problem249.py
''' Prime Subset Sums
Problem 249
Let S = {2, 3, 5, ..., 4999} be the set of prime numbers less than 5000.

Find the number of subsets of S, the sum of whose elements is a prime number.
Enter the rightmost 16 digits as your answer.
'''

import pdb
import pyprimes

'''
1548136 # desired sum of S
...
8 |
7 |
6 |
5 |  <8 billion entries>
4 |
3 |
2 |
1 |
0 |______________________
   1 2 3 4 5 6 7 ... 4999 # max usable prime (sUB)
'''

def takeWhile(p, xs):
    result = []
    for x in xs:
        if not p(x):
            break
        result.append(x)
    return result

sUB = 200 #5000

s = takeWhile(lambda p: p < sUB, pyprimes.primes())

hits = 0
misses = 0

# | Number of ways to form `n` as a summation of primes less than `maxPrime`
# (note the non-inclusiveness in the second parameter)
memo = dict()
# @profile
def numFormations(n, maxPrime):
    if (n == 0):
        return 1 # formation is the empty set

    global s
    global hits
    global misses

    cachedNumFormations = memo.get((n, maxPrime))
    if cachedNumFormations is not None:
        hits += 1
        return cachedNumFormations
    misses += 1

    result = 0
    for p in s:
        if p >= maxPrime or p > n:
            break
        result += numFormations(n - p, p)
        result %= 10000000000000000
    # print "f(", n, ",", maxPrime, ") = ", result
    memo[(n, maxPrime)] = result
    return result

# @profile
def main():
    global s
    global sUB
    global hits
    global misses

    print "S = takeWhile (< ", sUB, ") P.primes"
    print "Number of subsets of S that sum to a prime number: ",

    result = 0
    ssum = sum(s)
    maxUsablePrime = max(s)
    for p in pyprimes.primes():
        # `sum s` is for the subset that is `S`; the maximal subset
        if not p <= ssum:
            break

        # `+ 1` because of noninclusiveness
        result += numFormations(p, min(p + 1, maxUsablePrime + 1))
        result %= 10000000000000000
    print result
    # print 'cache size: ', len(memo)
    print 'hits: ', hits
    print 'misses: ', misses

if __name__ == '__main__':
    main()
